doc/sai-ptf/warmboot.md

```diff
@@ -0,0 +1,262 @@
+```
+@warm_setup
+@warm_test
+@warm_teardown
+```
+there is the sample code
+- Setup methods
+   ```python
+    @warm_setup
+    def setUp(self):
+        print("setUp WarmL2SanityTest")
+        L2SanityTest.setUp(self)
+
+
+    def setUp_starting(self):
+        print("setUp_starting WarmL2SanityTest")
+        SaiHelperBase.setUp(self)
+        super().param_init()
+
+
+    def setUp_post_start(self):
+        print("setUp_post_start WarmL2SanityTest")
+        SaiHelperBase.setUp(self)
+        super().param_init()
+   ```
+- runTest methods
+   ```python
+    @warm_test
+    def runTest(self):
+        print("Run test WarmL2SanityTest")
+        super().runTest()
+
+
+    def test_starting(self):
+        print("test_starting WarmL2SanityTest")
+        super().runTest()
+
+
+    def test_post_start(self):
+        print("test_post_start WarmL2SanityTest")
+        super().runTest()   
+   ```
+- tearDown Methods
+   ```python
+    @warm_teardown
+    def tearDown(self):
+        print("tearDown WarmL2SanityTest")
+        print("Skip the teardown and make a warm shut down for warm boot testing")
+        self.warm_shutdown()
+
+
+    def tearDown_starting(self):
+        print("tearDown_starting WarmL2SanityTest")
+        print("Skip the teardown for warm boot testing")
+
+
+    def tearDown_post_start(self):
+        print("tearDown_post_start WarmL2SanityTest")
+        print("Skip the teardown after warm boot testing")
+ ```
+
+### Code for extending the method in SAI-PTF
+There is the code logic for splitting the runTest method
+```python
+def warm_test(f):
+    """
+    Method decorator for the method of warm testing.
+    
+    Depends on parameters [test_reboot_mode] and [test_reboot_stage].
+    Runs different method, test_starting, setUp_post_start and runTest
+    """
+    def test_director(inst, *args):
+        if inst.test_reboot_stage == 'starting':
+            return inst.test_starting()
+        if inst.test_reboot_stage == 'post':
+            return inst.setUp_post_start()
+        return f(inst)
+    return test_director
+```
+
+## Support DUT different running stages during warm reboot
+During a warm reboot, we can split the DUT into three different stage
+- Before starting, in this stage, we will make some configurations or benchmark on the DUT, and use those setting to check the DUT functionality during a warm reboot. All the preparation for the warm reboot will be finished.
+- Starting, in this stage, DUT will run a warm-reboot command and not start any other docker services and sai-sdk will not start. We expect all the traffic will behave as normal.
+- After Starting, in this stage, sai-sdk will start, and data will be restored from the backup file. We expect there should be no change on the route and all the traffic will behave as normal. Meanwhile, if needed we can check the switch status from some SAI interfaces, as in this stage, SAI switch has been initialized and RPC APIs have been ready.
+
```

Comment:
It's better to move the above section closer to line 54 ("During a warm reboot, we need to split those three method in PTF-Test into three different stage...)

Reply:
ok, move this section upper to the section to introduce the different warm reboot stages